clang/lib/Headers/cheriintrin.h

```diff
@@ -0,0 +1,132 @@
+
+/* Object types, sealing and unsealing: */
+typedef long cheri_otype_t;
+#if defined(__mips__) || defined(__riscv)
+/* CHERI-MIPS and CHERI-RISCV use negative numbers for hardware-interpreted
+ * otypes */
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)-1)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)-2)
+#else
+#define CHERI_OTYPE_UNSEALED ((cheri_otype_t)0)
+#define CHERI_OTYPE_SENTRY ((cheri_otype_t)1)
+#endif
+#define cheri_gettype(x) __builtin_cheri_type_get(x)
+#define cheri_is_sealed(x) __builtin_cheri_sealed_get(x)
+#define cheri_is_sentry(x) (__builtin_cheri_type_get(x) == CHERI_OTYPE_SENTRY)
+#define cheri_is_unsealed(x) (!__builtin_cheri_sealed_get(x))
+/* TODO: builtins for indirect sentries */
+#define cheri_seal_entry(x) __builtin_cheri_seal_entry(x)
+#define cheri_seal(x, y) __builtin_cheri_seal((x), (y))
+#define cheri_unseal(x, y) __builtin_cheri_unseal((x), (y))
+
+/* Reconstruct capabilities from raw data: */
+#define cheri_build_cap(x, y) __builtin_cheri_cap_build((x), (y))
+#define cheri_conditional_seal(x, y) __builtin_cheri_conditional_seal((x), (y))
+#define cheri_copy_type(x, y) __builtin_cheri_cap_type_copy((x), (y))
+
+/* Capability permissions: */
+enum __attribute__((flag_enum, enum_extensibility(open))) CheriPermissions {
+  CHERI_PERM_GLOBAL = __CHERI_CAP_PERMISSION_GLOBAL__,
+  CHERI_PERM_EXECUTE = __CHERI_CAP_PERMISSION_PERMIT_EXECUTE__,
+  CHERI_PERM_LOAD = __CHERI_CAP_PERMISSION_PERMIT_LOAD__,
+  CHERI_PERM_STORE = __CHERI_CAP_PERMISSION_PERMIT_STORE__,
+  CHERI_PERM_LOAD_CAP = __CHERI_CAP_PERMISSION_PERMIT_LOAD_CAPABILITY__,
+  CHERI_PERM_STORE_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_CAPABILITY__,
+  CHERI_PERM_STORE_LOCAL_CAP = __CHERI_CAP_PERMISSION_PERMIT_STORE_LOCAL__,
+  CHERI_PERM_SEAL = __CHERI_CAP_PERMISSION_PERMIT_SEAL__,
+  CHERI_PERM_CCALL = __CHERI_CAP_PERMISSION_PERMIT_CCALL__,
+  CHERI_PERM_UNSEAL = __CHERI_CAP_PERMISSION_PERMIT_UNSEAL__,
+  CHERI_PERM_SYSTEM_REGS = __CHERI_CAP_PERMISSION_ACCESS_SYSTEM_REGISTERS__,
+  /* TODO: architecture-dependent permissions */
+};
+typedef enum CheriPermissions cheri_perms_t;
+#define cheri_getperms(x) ((cheri_perms_t)__builtin_cheri_perms_get(x))
+#define cheri_andperms(x, y) __builtin_cheri_perms_and((x), (cheri_perms_t)(y))
+#define cheri_clearperms(x, y) cheri_andperms((x), ~(y))
+
+/* Accessors for capability registers. Currently exposes DDC,PCC and stack. */
+#define cheri_getddc() __builtin_cheri_global_data_get()
+#define cheri_getpcc() __builtin_cheri_program_counter_get()
+#define cheri_getstack() __builtin_cheri_stack_get()
+
+/* Partially portable builtins: */
+/* Note: {get,set}flags does nothing for MIPS, but can still be used. */
+#define cheri_getflags(x) __builtin_cheri_flags_get(x)
+#define cheri_setflags(x, y) __builtin_cheri_flags_set((x), (y))
+#define cheri_load_tags(x) __builtin_cheri_cap_load_tags(x)
+
+/*
+ * Alignment builtins: Not CHERI-specific, but motivated by CHERI.
```

Comment:
Should we encourage direct use of the `__builtin` rather than making them seem CHERI-specific?

Reply:
Yes, that would be ideal, but it's quite annoying to type.
I wonder if we could provide something like 
`_Align_{up,down}()`/`_Is_aligned()`/`___align_{up,down}()`/`__is_aligned()` in stdalign.h. 
Since those are reserved identifiers that should be fine and a bit shorter than __builtin_align_up, etc.

Reply:
align_up, etc seem to already be used in lots of projects: https://github.com/search?l=C&q=align_up&type=Code

Reply:
In the short term, having providing easy-to-use CHERI-specific macros is at least very useful to us.  We can alway migrate to a newer, better portable thing in the future, should one arise.